flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/PatternStream.scala

```diff
@@ -0,0 +1,142 @@
+
+/**
+  * Stream abstraction for CEP pattern detection. A pattern stream is a stream which emits detected
+  * pattern sequences as a map of events associated with their names. The pattern is detected using a
+  * { @link org.apache.flink.cep.nfa.NFA}. In order to process the detected sequences, the user
+  * has to specify a { @link PatternSelectFunction} or a { @link PatternFlatSelectFunction}.
+  *
+  * @param jPatternStream Underlying pattern stream from Java API
+  * @tparam T Type of the events
+  */
+class PatternStream[T: TypeInformation : ClassTag](jPatternStream: JPatternStream[T]) {
+
+  private[flink] def getWrappedPatternStream = jPatternStream
+
+  /**
+    * Applies a select function to the detected pattern sequence. For each pattern sequence the
+    * provided { @link PatternSelectFunction} is called. The pattern select function can produce
+    * exactly one resulting element.
+    *
+    * @param patternSelectFunction The pattern select function which is called for each detected
+    *                              pattern sequence.
+    * @tparam R Type of the resulting elements
+    * @return { @link DataStream} which contains the resulting elements from the pattern select
+    *         unction.
+    */
+  def select[R: TypeInformation : ClassTag](patternSelectFunction: PatternSelectFunction[T, R]): DataStream[R] = {
+    asScalaStream(jPatternStream.select(patternSelectFunction, implicitly[TypeInformation[R]]))
+  }
+
+  /**
+    * Applies a flat select function to the detected pattern sequence. For each pattern sequence
+    * the provided { @link PatternFlatSelectFunction} is called. The pattern flat select function
+    * can produce an arbitrary number of resulting elements.
+    *
+    * @param patternFlatSelectFunction The pattern flat select function which is called for each
+    *                                  detected pattern sequence.
+    * @tparam R Type of the resulting elements
+    * @return { @link DataStream} which contains the resulting elements from the pattern flat select
+    *         function.
+    */
+  def flatSelect[R: TypeInformation : ClassTag](patternFlatSelectFunction: PatternFlatSelectFunction[T, R]): DataStream[R] = {
+    asScalaStream(jPatternStream.flatSelect(patternFlatSelectFunction, implicitly[TypeInformation[R]]))
+  }
+
+  /**
+    * Applies a select function to the detected pattern sequence. For each pattern sequence the
+    * provided { @link PatternSelectFunction} is called. The pattern select function can produce
+    * exactly one resulting element.
+    *
+    * @param patternSelectFun The pattern select function which is called for each detected
+    *                         pattern sequence.
+    * @tparam R Type of the resulting elements
+    * @return { @link DataStream} which contains the resulting elements from the pattern select
+    *         function.
+    */
+  def select[R: TypeInformation : ClassTag](patternSelectFun: JMap[String, T] => R): DataStream[R] = {
```

Comment:
I think it would be good to not give a Java `Map` but a Scala `Map` to the `patternSelectFun` as you've done in the out commented code.

I think it would be good to use an explicit conversion via the `JavaConverters` and the `asScala` method.

Semantically, it would be correct to provide a immutable map to the Scala lambda. However, this would mean that we copy the map. Therefore, for the sake of efficiency, we could simply pass in the result of the `asScala` call which is a mutable `Map`.